src/Exception/HttpExceptionTrait.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace Superbrave\VerboseErrorHttpClient\Exception;
+
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Throwable;
+
+/**
+ * @author Niels Nijens <nn@superbrave.nl>
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @internal
+ */
+trait HttpExceptionTrait
+{
+    /**
+     * @var ResponseInterface
+     */
+    private $response;
+
+    /**
+     * @param ResponseInterface $response
+     * @param Throwable         $previousException
+     */
+    public function __construct(ResponseInterface $response, Throwable $previousException)
+    {
+        $this->response = $response;
+
+        $code = $response->getInfo('http_code');
+        $url = $response->getInfo('url');
+        $message = sprintf('HTTP %d returned for "%s".', $code, $url);
+
+        $httpCodeFound = false;
+        $isJson = false;
+        foreach (array_reverse($response->getInfo('response_headers')) as $h) {
+            if (0 === strpos($h, 'HTTP/')) {
+                if ($httpCodeFound) {
+                    break;
+                }
+
+                $message = sprintf('%s returned for "%s".', $h, $url);
+                $httpCodeFound = true;
+            }
+
+            if (0 === stripos($h, 'content-type:')) {
+                if (preg_match('/\bjson\b/i', $h)) {
+                    $isJson = true;
+                }
+
+                if ($httpCodeFound) {
+                    break;
+                }
+            }
+        }
+
+        $contentSummary = $this->getResponseContentSummary($isJson);
+        if (empty($contentSummary) === false) {
+            $message = sprintf('%s with response: %s', rtrim($message, '.'), $contentSummary);
+        }
+
+        parent::__construct($message, $code, $previousException);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getResponse(): ResponseInterface
+    {
+        return $this->response;
+    }
+
+    /**
+     * Returns a summary of the response content.
+     *
+     * @param bool $json
+     *
+     * @return string
+     */
+    private function getResponseContentSummary(bool $json): string
+    {
+        $content = $this->response->getContent(false);
+        if ($json) {
+            return $content;
+        }
+
+        $contentSummary = substr($content, 0, 120);
```

Comment:
Why truncate at all? Wouldn't it be useful to have the entire response? 120 chars is pretty restrictive to have for error messages.

Reply:
The response body is only truncated when it's HTML or long plain text. When appending an entire HTML response this will most likely become unreadable in error logging system. When a JSON body is returned (the most likely response for APIs) it is entirely added to the exception message. The 120 character length was inspired by Guzzle's exception message implementation.